backend/lucene/src/main/java/org/hibernate/search/backend/lucene/lowlevel/query/impl/VectorSimilarityFilterQuery.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.VectorUtil;
+
+public class VectorSimilarityFilterQuery extends Query {
+
+	private final Query query;
+	private final float similarityAsScore;
+
+	public static VectorSimilarityFilterQuery create(KnnByteVectorQuery query, float similarityLimit,
+			VectorSimilarityFunction vectorSimilarityFunction) {
+		// We assume that `similarityLimit` is a distance so we need to convert it to the score using a formula from a
+		// similarity function:
+		return new VectorSimilarityFilterQuery(
+				query, byteSimilarityDistanceToScore( similarityLimit, vectorSimilarityFunction ) );
+	}
+
+	public static VectorSimilarityFilterQuery create(KnnFloatVectorQuery query, float similarityLimit,
+			VectorSimilarityFunction vectorSimilarityFunction) {
+		// We assume that `similarityLimit` is a distance so we need to convert it to the score using a formula from a
+		// similarity function:
+		return new VectorSimilarityFilterQuery(
+				query, floatSimilarityDistanceToScore( similarityLimit, vectorSimilarityFunction ) );
+	}
+
+	private VectorSimilarityFilterQuery(Query query, float similarityAsScore) {
+		this.query = query;
+		this.similarityAsScore = similarityAsScore;
+	}
+
+	@Override
+	public Query rewrite(IndexSearcher indexSearcher) throws IOException {
+		Query rewritten = query.rewrite( indexSearcher );
+		if ( rewritten == query ) {
+			return this;
+		}
+		// Knn queries are rewritten and we need to use a rewritten one to get the weights and scores:
+		return new VectorSimilarityFilterQuery( rewritten, this.similarityAsScore );
+	}
+
+	@Override
+	public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+		// we've already converted distance/similarity to a score, but now if the underlying query is boosting the score,
+		// we'd want to boost our converted one as well to get the expected matches:
+		return new SimilarityWeight( query.createWeight( searcher, scoreMode, boost ), similarityAsScore * boost );
+	}
+
+	@Override
+	public void visit(QueryVisitor visitor) {
+		visitor.visitLeaf( this );
+	}
+
+	@Override
+	public String toString(String field) {
+		return getClass().getName() + "{" +
+				"query=" + query +
+				", similarityLimit=" + similarityAsScore +
+				'}';
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if ( this == o ) {
+			return true;
+		}
+		if ( o == null || getClass() != o.getClass() ) {
+			return false;
+		}
+		VectorSimilarityFilterQuery that = (VectorSimilarityFilterQuery) o;
+		return Float.compare( similarityAsScore, that.similarityAsScore ) == 0 && Objects.equals( query, that.query );
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash( query, similarityAsScore );
+	}
+
+	private static float floatSimilarityDistanceToScore(float distance, VectorSimilarityFunction similarityFunction) {
+		switch ( similarityFunction ) {
+			case EUCLIDEAN:
+				return 1.0f / ( 1.0f + distance );
```

Comment:
NOTE: from what it seems in Elasticsearch a different formula is used to convert the similarity into score:
```
l2_norm: sqrt((1 / _score) - 1) 
```
from which score is:
```
l2_norm:  _score = (1 / d^2+1) 
```
but then in a Lucene function: `1 / (1 + squareDistance(v1, v2))`